File: sdks/node/src/meta-llm/types/usage.ts
/**
 * ADR-0024b §D3's `UsageSummary`/`BudgetView`, plus the bounded query the
 * read-only `client.usage()` method (`../client.ts`) accepts.
 *
 * Usage is strictly authenticated-account scoped (§D3) — every query is
 * bound to the caller's own credential; there is no cross-tenant or
 * cross-account parameter anywhere in {@link UsageQuery}. An empty
 * `UsageSummary` is not reinterpreted as "no usage anywhere" vs "this
 * account genuinely has none" (§D3) — `client.usage()` returns whatever
 * the server reports as-is, with no speculative fallback logic layered on
 * top.
 */

import type { Money } from "./money.js";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export interface CacheStats {
  hitRate?: number;
  savings?: Money;
  raw?: Record<string, unknown>;
}

export interface UsageTotals {
  requests?: number;
  promptTokens?: number;
  completionTokens?: number;
  totalTokens?: number;
  cost?: Money;
  raw?: Record<string, unknown>;
}

/**
 * Plan degradation and reset information are preserved as reported (§D4) —
 * this SDK never recomputes `status`/`headroom` from the other fields.
 */
export interface BudgetView {
  serving?: Money;
  hardLimit?: Money;
  committed?: Money;
  reserved?: Money;
  headroom?: Money;
  status?: string;
  resetsAt?: string;
  raw?: Record<string, unknown>;
}

export interface UsageBreakdownEntry {
  requests?: number;
  cost?: Money;
  raw?: Record<string, unknown>;
}

export interface UsagePeriodEntry extends UsageBreakdownEntry {
  period: string;
}

/** ADR-0024b §D3's `UsageSummary`. */
export interface UsageSummary {
  totals: UsageTotals;
  tierMix?: Record<string, number>;
  escalationRate?: number;
  cache?: CacheStats;
  fallbackRate?: number;
  emptyBilledRate?: number;
  byModel?: Record<string, UsageBreakdownEntry>;
  byProvider?: Record<string, UsageBreakdownEntry>;
  byPeriod?: UsagePeriodEntry[];
  budget?: BudgetView;
  /** Fields present on the wire this decoder does not recognize, preserved verbatim (never dropped). */
  raw?: Record<string, unknown>;
}

/** Bounded `YYYY-MM` query window plus optional grouping (ADR-0024b §D3). */
export interface UsageQuery {
  /** Inclusive `YYYY-MM` start of the query range. */
  from: string;
  /** Inclusive `YYYY-MM` end of the query range. */
  to: string;
  model?: string;
  provider?: string;
  groupBy?: "model" | "provider" | "period";
}

const YYYY_MM = /^\d{4}-(0[1-9]|1[0-2])$/;

export class InvalidUsageQueryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidUsageQueryError";
  }
}

/** Validates the bounded `YYYY-MM` range required by §D3 before any request is sent. */
export function assertValidUsageQuery(query: UsageQuery): void {
  if (!YYYY_MM.test(query.from)) {
    throw new InvalidUsageQueryError(`UsageQuery.from must match YYYY-MM; got ${JSON.stringify(query.from)}`);
  }
  if (!YYYY_MM.test(query.to)) {
    throw new InvalidUsageQueryError(`UsageQuery.to must match YYYY-MM; got ${JSON.stringify(query.to)}`);
  }
  if (query.from > query.to) {
    throw new InvalidUsageQueryError(
      `UsageQuery.from (${JSON.stringify(query.from)}) must not be after .to (${JSON.stringify(query.to)})`,
    );
  }
}

function parseCacheStats(raw: unknown): CacheStats | undefined {
  if (!isRecord(raw)) return undefined;
  const known = new Set(["hit_rate", "hitRate", "savings"]);
  const rawRemainder: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(raw)) {
    if (!known.has(key)) rawRemainder[key] = value;
  }
  const hitRateRaw = raw.hit_rate ?? raw.hitRate;
  return {
    hitRate: typeof hitRateRaw === "number" ? hitRateRaw : undefined,
    savings: parseMoneyImport(raw.savings),
    raw: Object.keys(rawRemainder).length > 0 ? rawRemainder : undefined,
  };
}

function parseUsageTotals(raw: unknown): UsageTotals {
  if (!isRecord(raw)) return {};
  const known = new Set([
    "requests",
    "prompt_tokens",
    "promptTokens",
    "completion_tokens",
    "completionTokens",
    "total_tokens",
    "totalTokens",
    "cost",
  ]);
  const rawRemainder: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(raw)) {
    if (!known.has(key)) rawRemainder[key] = value;
  }
  return {
    requests: typeof raw.requests === "number" ? raw.requests : undefined,
    promptTokens: typeof (raw.prompt_tokens ?? raw.promptTokens) === "number" ? (raw.prompt_tokens as number) ?? (raw.promptTokens as number) : undefined,
    completionTokens:
      typeof (raw.completion_tokens ?? raw.completionTokens) === "number"
        ? (raw.completion_tokens as number) ?? (raw.completionTokens as number)
        : undefined,
    totalTokens:
      typeof (raw.total_tokens ?? raw.totalTokens) === "number"
        ? (raw.total_tokens as number) ?? (raw.totalTokens as number)
        : undefined,
    cost: parseMoneyImport(raw.cost),
    raw: Object.keys(rawRemainder).length > 0 ? rawRemainder : undefined,
  };
}

function parseBudgetView(raw: unknown): BudgetView | undefined {
  if (!isRecord(raw)) return undefined;
  const known = new Set([
    "serving",
    "hard_limit",
    "hardLimit",
    "committed",
    "reserved",
    "headroom",
    "status",
    "resets_at",
    "resetsAt",
  ]);
  const rawRemainder: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(raw)) {
    if (!known.has(key)) rawRemainder[key] = value;
  }
  const resetsAtRaw = raw.resets_at ?? raw.resetsAt;
  return {
    serving: parseMoneyImport(raw.serving),
    hardLimit: parseMoneyImport(raw.hard_limit ?? raw.hardLimit),
    committed: parseMoneyImport(raw.committed),
    reserved: parseMoneyImport(raw.reserved),
    headroom: parseMoneyImport(raw.headroom),
    status: typeof raw.status === "string" ? raw.status : undefined,
    resetsAt: typeof resetsAtRaw === "string" ? resetsAtRaw : undefined,
    raw: Object.keys(rawRemainder).length > 0 ? rawRemainder : undefined,
  };
}

function parseBreakdownEntry(raw: unknown): UsageBreakdownEntry {
  if (!isRecord(raw)) return {};
  const known = new Set(["requests", "cost"]);
  const rawRemainder: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(raw)) {
    if (!known.has(key)) rawRemainder[key] = value;
  }
  return {
    requests: typeof raw.requests === "number" ? raw.requests : undefined,
    cost: parseMoneyImport(raw.cost),
    raw: Object.keys(rawRemainder).length > 0 ? rawRemainder : undefined,
  };
}

function parseBreakdownMap(raw: unknown): Record<string, UsageBreakdownEntry> | undefined {
  if (!isRecord(raw)) return undefined;
  const out: Record<string, UsageBreakdownEntry> = {};
  for (const [key, value] of Object.entries(raw)) {
    out[key] = parseBreakdownEntry(value);
  }
  return out;
}

function parsePeriodEntries(raw: unknown): UsagePeriodEntry[] | undefined {
  if (!Array.isArray(raw)) return undefined;
  return raw
    .filter((item): item is Record<string, unknown> => isRecord(item) && typeof item.period === "string")
    .map((item) => ({ period: item.period as string, ...parseBreakdownEntry(item) }));
}

// Re-declared locally to avoid a second import line churning the diff if
// `./money.js`'s export surface changes; keeps this module's parse helpers
// self-contained.
function parseMoneyImport(raw: unknown): Money | undefined {
  if (!isRecord(raw)) return undefined;
  const amountRaw = raw.amount;
  const currency = raw.currency ?? raw.currency_code ?? raw.currencyCode;
  if ((typeof amountRaw === "string" || typeof amountRaw === "number") && typeof currency === "string") {
    return { amount: String(amountRaw), currency };
  }
  return undefined;
}

const KNOWN_USAGE_KEYS = new Set([
  "totals",
  "tier_mix",
  "tierMix",
  "escalation_rate",
  "escalationRate",
  "cache",
  "fallback_rate",
  "fallbackRate",
  "empty_billed_rate",
  "emptyBilledRate",
  "by_model",
  "byModel",
  "by_provider",
  "byProvider",
  "by_period",
  "byPeriod",
  "budget",
]);

/**
 * Parse a raw `/v1/usage` JSON body into a typed {@link UsageSummary}.
 * Never throws — an entirely empty/malformed body decodes to an
 * `UsageSummary` with empty `totals` rather than an error, since an empty
 * result is itself meaningful account-scoped evidence (§D3), not a parse
 * failure.
 */
export function parseUsageSummary(raw: unknown): UsageSummary {
  if (!isRecord(raw)) return { totals: {} };

  const rawRemainder: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(raw)) {
    if (!KNOWN_USAGE_KEYS.has(key)) rawRemainder[key] = value;
  }

  const escalationRateRaw = raw.escalation_rate ?? raw.escalationRate;
  const fallbackRateRaw = raw.fallback_rate ?? raw.fallbackRate;
  const emptyBilledRateRaw = raw.empty_billed_rate ?? raw.emptyBilledRate;
  const tierMixRaw = raw.tier_mix ?? raw.tierMix;

  return {
    totals: parseUsageTotals(raw.totals),
    tierMix: isRecord(tierMixRaw) ? (tierMixRaw as Record<string, number>) : undefined,
    escalationRate: typeof escalationRateRaw === "number" ? escalationRateRaw : undefined,
    cache: parseCacheStats(raw.cache),
    fallbackRate: typeof fallbackRateRaw === "number" ? fallbackRateRaw : undefined,
    emptyBilledRate: typeof emptyBilledRateRaw === "number" ? emptyBilledRateRaw : undefined,
    byModel: parseBreakdownMap(raw.by_model ?? raw.byModel),
    byProvider: parseBreakdownMap(raw.by_provider ?? raw.byProvider),
    byPeriod: parsePeriodEntries(raw.by_period ?? raw.byPeriod),
    budget: parseBudgetView(raw.budget),
    raw: Object.keys(rawRemainder).length > 0 ? rawRemainder : undefined,
  };
}

File: sdks/rust/src/meta_llm/types/usage.rs
//! ADR-0024b §D3's `UsageSummary`/`BudgetView`, plus the bounded query the
//! read-only `MetaLlmClient::usage` method (`super::super::client`)
//! accepts.
//!
//! Usage is strictly authenticated-account scoped (§D3) -- every query is
//! bound to the caller's own credential; there is no cross-tenant or
//! cross-account parameter anywhere in [`UsageQuery`]. An empty
//! `UsageSummary` is not reinterpreted as "no usage anywhere" vs. "this
//! account genuinely has none" (§D3) -- `usage()` returns whatever the
//! server reports as-is, with no speculative fallback logic layered on
//! top.

use std::collections::HashMap;

use serde_json::{Map, Value};

use super::money::{parse_money, Money};

#[derive(Debug, Clone, Default, PartialEq)]
pub struct CacheStats {
    pub hit_rate: Option<f64>,
    pub savings: Option<Money>,
    pub raw: Option<Map<String, Value>>,
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct UsageTotals {
    pub requests: Option<u64>,
    pub prompt_tokens: Option<u64>,
    pub completion_tokens: Option<u64>,
    pub total_tokens: Option<u64>,
    pub cost: Option<Money>,
    pub raw: Option<Map<String, Value>>,
}

/// Plan degradation and reset information are preserved as reported
/// (§D4) -- this SDK never recomputes `status`/`headroom` from the other
/// fields.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct BudgetView {
    pub serving: Option<Money>,
    pub hard_limit: Option<Money>,
    pub committed: Option<Money>,
    pub reserved: Option<Money>,
    pub headroom: Option<Money>,
    pub status: Option<String>,
    pub resets_at: Option<String>,
    pub raw: Option<Map<String, Value>>,
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct UsageBreakdownEntry {
    pub requests: Option<u64>,
    pub cost: Option<Money>,
    pub raw: Option<Map<String, Value>>,
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct UsagePeriodEntry {
    pub period: String,
    pub requests: Option<u64>,
    pub cost: Option<Money>,
    pub raw: Option<Map<String, Value>>,
}

/// ADR-0024b §D3's `UsageSummary`.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct UsageSummary {
    pub totals: UsageTotals,
    pub tier_mix: Option<HashMap<String, f64>>,
    pub escalation_rate: Option<f64>,
    pub cache: Option<CacheStats>,
    pub fallback_rate: Option<f64>,
    pub empty_billed_rate: Option<f64>,
    pub by_model: Option<HashMap<String, UsageBreakdownEntry>>,
    pub by_provider: Option<HashMap<String, UsageBreakdownEntry>>,
    pub by_period: Option<Vec<UsagePeriodEntry>>,
    pub budget: Option<BudgetView>,
    /// Fields present on the wire this decoder does not recognize, preserved verbatim (never dropped).
    pub raw: Option<Map<String, Value>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UsageGroupBy {
    Model,
    Provider,
    Period,
}

impl UsageGroupBy {
    pub fn as_query_str(self) -> &'static str {
        match self {
            UsageGroupBy::Model => "model",
            UsageGroupBy::Provider => "provider",
            UsageGroupBy::Period => "period",
        }
    }
}

/// Bounded `YYYY-MM` query window plus optional grouping (ADR-0024b §D3).
#[derive(Debug, Clone)]
pub struct UsageQuery {
    /// Inclusive `YYYY-MM` start of the query range.
    pub from: String,
    /// Inclusive `YYYY-MM` end of the query range.
    pub to: String,
    pub model: Option<String>,
    pub provider: Option<String>,
    pub group_by: Option<UsageGroupBy>,
}

impl UsageQuery {
    pub fn new(from: impl Into<String>, to: impl Into<String>) -> Self {
        Self {
            from: from.into(),
            to: to.into(),
            model: None,
            provider: None,
            group_by: None,
        }
    }
}

#[derive(Debug, Clone, thiserror::Error)]
#[error("{0}")]
pub struct InvalidUsageQueryError(pub String);

fn is_valid_yyyy_mm(s: &str) -> bool {
    let bytes = s.as_bytes();
    if bytes.len() != 7 {
        return false;
    }
    if !bytes[..4].iter().all(u8::is_ascii_digit) {
        return false;
    }
    if bytes[4] != b'-' {
        return false;
    }
    if !bytes[5..7].iter().all(u8::is_ascii_digit) {
        return false;
    }
    s[5..7].parse::<u8>().map(|m| (1..=12).contains(&m)).unwrap_or(false)
}

/// Validates the bounded `YYYY-MM` range required by §D3 before any request is sent.
pub fn assert_valid_usage_query(query: &UsageQuery) -> Result<(), InvalidUsageQueryError> {
    if !is_valid_yyyy_mm(&query.from) {
        return Err(InvalidUsageQueryError(format!(
            "UsageQuery.from must match YYYY-MM; got {:?}",
            query.from
        )));
    }
    if !is_valid_yyyy_mm(&query.to) {
        return Err(InvalidUsageQueryError(format!(
            "UsageQuery.to must match YYYY-MM; got {:?}",
            query.to
        )));
    }
    if query.from > query.to {
        return Err(InvalidUsageQueryError(format!(
            "UsageQuery.from ({:?}) must not be after .to ({:?})",
            query.from, query.to
        )));
    }
    Ok(())
}

fn num_field(obj: &Map<String, Value>, key: &str) -> Option<f64> {
    obj.get(key).and_then(Value::as_f64)
}

fn u64_field(obj: &Map<String, Value>, key: &str) -> Option<u64> {
    obj.get(key).and_then(Value::as_u64)
}

fn raw_remainder(obj: &Map<String, Value>, known: &[&str]) -> Option<Map<String, Value>> {
    let out: Map<String, Value> = obj
        .iter()
        .filter(|(k, _)| !known.contains(&k.as_str()))
        .map(|(k, v)| (k.clone(), v.clone()))
        .collect();
    if out.is_empty() {
        None
    } else {
        Some(out)
    }
}

fn parse_cache_stats(raw: &Value) -> Option<CacheStats> {
    let obj = raw.as_object()?;
    let known = ["hit_rate", "savings"];
    Some(CacheStats {
        hit_rate: num_field(obj, "hit_rate"),
        savings: obj.get("savings").and_then(parse_money),
        raw: raw_remainder(obj, &known),
    })
}

fn parse_usage_totals(raw: Option<&Value>) -> UsageTotals {
    let Some(obj) = raw.and_then(Value::as_object) else {
        return UsageTotals::default();
    };
    let known = [
        "requests",
        "prompt_tokens",
        "completion_tokens",
        "total_tokens",
        "cost",
    ];
    UsageTotals {
        requests: u64_field(obj, "requests"),
        prompt_tokens: u64_field(obj, "prompt_tokens"),
        completion_tokens: u64_field(obj, "completion_tokens"),
        total_tokens: u64_field(obj, "total_tokens"),
        cost: obj.get("cost").and_then(parse_money),
        raw: raw_remainder(obj, &known),
    }
}

fn parse_budget_view(raw: &Value) -> Option<BudgetView> {
    let obj = raw.as_object()?;
    let known = [
        "serving",
        "hard_limit",
        "committed",
        "reserved",
        "headroom",
        "status",
        "resets_at",
    ];
    Some(BudgetView {
        serving: obj.get("serving").and_then(parse_money),
        hard_limit: obj.get("hard_limit").and_then(parse_money),
        committed: obj.get("committed").and_then(parse_money),
        reserved: obj.get("reserved").and_then(parse_money),
        headroom: obj.get("headroom").and_then(parse_money),
        status: obj.get("status").and_then(|v| v.as_str()).map(str::to_owned),
        resets_at: obj.get("resets_at").and_then(|v| v.as_str()).map(str::to_owned),
        raw: raw_remainder(obj, &known),
    })
}

fn parse_breakdown_entry(raw: &Value) -> UsageBreakdownEntry {
    let Some(obj) = raw.as_object() else {
        return UsageBreakdownEntry::default();
    };
    let known = ["requests", "cost"];
    UsageBreakdownEntry {
        requests: u64_field(obj, "requests"),
        cost: obj.get("cost").and_then(parse_money),
        raw: raw_remainder(obj, &known),
    }
}

fn parse_breakdown_map(raw: Option<&Value>) -> Option<HashMap<String, UsageBreakdownEntry>> {
    let obj = raw.and_then(Value::as_object)?;
    Some(
        obj.iter()
            .map(|(k, v)| (k.clone(), parse_breakdown_entry(v)))
            .collect(),
    )
}

fn parse_period_entries(raw: Option<&Value>) -> Option<Vec<UsagePeriodEntry>> {
    let arr = raw.and_then(Value::as_array)?;
    Some(
        arr.iter()
            .filter_map(|item| {
                let period = item.as_object()?.get("period")?.as_str()?.to_owned();
                let entry = parse_breakdown_entry(item);
                Some(UsagePeriodEntry {
                    period,
                    requests: entry.requests,
                    cost: entry.cost,
                    raw: entry.raw,
                })
            })
            .collect(),
    )
}

const KNOWN_USAGE_KEYS: &[&str] = &[
    "totals",
    "tier_mix",
    "escalation_rate",
    "cache",
    "fallback_rate",
    "empty_billed_rate",
    "by_model",
    "by_provider",
    "by_period",
    "budget",
];

/// Parse a raw `/v1/usage` JSON body into a typed [`UsageSummary`]. Never
/// panics -- an entirely empty/malformed body decodes to a `UsageSummary`
/// with empty `totals` rather than an error, since an empty result is
/// itself meaningful account-scoped evidence (§D3), not a parse failure.
pub fn parse_usage_summary(raw: &Value) -> UsageSummary {
    let Some(obj) = raw.as_object() else {
        return UsageSummary::default();
    };

    let tier_mix = obj.get("tier_mix").and_then(Value::as_object).map(|m| {
        m.iter()
            .filter_map(|(k, v)| v.as_f64().map(|n| (k.clone(), n)))
            .collect()
    });

    UsageSummary {
        totals: parse_usage_totals(obj.get("totals")),
        tier_mix,
        escalation_rate: num_field(obj, "escalation_rate"),
        cache: obj.get("cache").and_then(parse_cache_stats),
        fallback_rate: num_field(obj, "fallback_rate"),
        empty_billed_rate: num_field(obj, "empty_billed_rate"),
        by_model: parse_breakdown_map(obj.get("by_model")),
        by_provider: parse_breakdown_map(obj.get("by_provider")),
        by_period: parse_period_entries(obj.get("by_period")),
        budget: obj.get("budget").and_then(parse_budget_view),
        raw: raw_remainder(obj, KNOWN_USAGE_KEYS),
    }
}
